### PR TITLE
refactor(encode): optimize trie, pick shortest entities, use base-10 numerics

### PR DESCRIPTION
**Pull request by fb55**
## Summary

- **Pick shortest entity on encode**: the encode-trie generator now selects the shortest named entity when multiple aliases exist for the same code point (e.g. `&amp;` over `&AMP;`)
- **Base-10 numeric entities**: switches fallback numeric encoding from `&#xHEX;` to `&#DEC;` — shorter output for most code points (e.g. `&#10084;` vs `&#x2764;`)
- **Trie overhaul**: replaces the imported `htmlTrie` Map from `encode-html.ts` with a compact serialized string format parsed at startup into a `Map<number, EncodeTrieNode>` + a flat `asciiEntities[]` lookup table for fast ASCII paths
- **Deleted `src/internal/encode-shared.ts`**: all encode trie logic now lives in `encode.ts`
- **Multi-code-point entities preserved**: entities like `&nvlt;` (`<` + U+20D2), `&nvgt;` (`>` + U+20D2), and `&bne;` (`=` + U+20E5) are kept in the trie and the encoder checks for multi-char matches before the ASCII fast path

## Breaking changes

- **Numeric entity output changes from hex to decimal** — `&#x2764;` becomes `&#10084;`. Both are valid HTML/XML, but snapshot tests matching exact output will need updating.
- **Named entity output may change** when a shorter alias exists — e.g. if both `&AMP;` and `&amp;` map to the same code point, the shorter one is now always chosen.
- The `escapeUTF8` / `encodeXML` functions in `escape.ts` also produce decimal numeric entities (`&#...;` instead of `&#x...;`).
- `src/internal/encode-shared.ts` is removed.

## Benchmark results

Geometric mean across 12 text variants (short/medium/long × low/high entity density × named/numeric), via [html-entity-benchmarks](https://github.com/nicolo-ribaudo/html-entity-benchmarks) `compare`:

| | Geo mean (ops/s) | Δ |
|---|---|---|
| **entities v8** (published) | 541.5K | — |
| **encode-perf** | 582.9K | **+8%** |

Biggest gain on short-low-named inputs (2.94M → 7.65M, **+2.6×**) from the ASCII fast-path array lookup. Longer/higher-density inputs are roughly at parity.

## Bundle size (esbuild --minify --tree-shaking, esm)

| Export | Before | After | Δ |
|---|---|---|---|
| `encodeHTML` | 14,455 B (6,242 gz) | 11,150 B (5,328 gz) | **−23%** (−15% gz) |
| Full bundle | 51,718 B (28,091 gz) | 48,605 B (27,320 gz) | −6% (−3% gz) |

🤖 Generated with Claude Opus ([Claude Code](https://claude.com/claude-code))

## Summary by CodeRabbit

* **Refactor**
  * Numeric character references now use decimal form instead of hexadecimal.
* **Bug Fixes**
  * Improved handling of full Unicode (including astral/surrogate pairs) and multi-code-point sequences so emoji and combined characters encode correctly.
* **Tests**
  * Updated and added tests to expect decimal references and to cover multi-code-point/edge-case sequences.

### CODE REVIEW

**Comment by fb55** [comment]
@coderabbitai review

File: src/encode.ts [hunk 1]
```diff
@@ -1,18 +1,152 @@
...
+            data.charCodeAt(cursor) === 123 /* '{' */ ? null : readEntity();
+
+        if (data.charCodeAt(cursor) === 123 /* '{' */) {
+            cursor++; // Skip '{'
+            const next = new Map<number, EncodeTrieNode>();
+            let childKey = -1;
+            while (data.charCodeAt(cursor) !== 125 /* '}' */) {
+                childKey += readGap() + 1;
+                next.set(childKey, readEntity());
+            }
```
[the remark this answers is not in the document]
Comment (fb55) [reply src/encode.ts:91]: Not relevant
